src/ngx_http_graphite_module.c

```diff
@@ -1794,14 +1815,78 @@ ngx_http_graphite_timer_event_handler(ngx_event_t *ev) {
         char *next = NULL;
         char *nl = NULL;
 
-        int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+        int fd;
+         if (ngx_strcmp(gmcf->protocol.data, "tcp") == 0) {
+             fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+         } else {
+             fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+         }
+
         if (fd < 0) {
             ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "graphite can't create socket");
             if (!(ngx_quit || ngx_terminate || ngx_exiting))
                 ngx_add_timer(ev, gmcf->frequency);
             return;
         }
 
+        if (ngx_strcmp(gmcf->protocol.data, "tcp") == 0) {
+
+            int flags;
+            fd_set myset;
+
+            /* Set socket to non-blocking */
+            if ((flags = fcntl(fd, F_GETFL, 0)) < 0)
+            {
+                ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "failed to get the current non-blocking flag");
+                if (!(ngx_quit || ngx_terminate || ngx_exiting))
+                    ngx_add_timer(ev, gmcf->frequency);
+                return;
+            }
+
+            if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0)
+            {
+                ngx_log_error(NGX_LOG_ALERT, ev->log, 0, "failed to set the socket non-blocking");
+                if (!(ngx_quit || ngx_terminate || ngx_exiting))
+                    ngx_add_timer(ev, gmcf->frequency);
+                return;
+            }
+              int res = connect(fd, (struct sockaddr *)&sin, sizeof(sin));
+              if (res < 0) {
+                 if (errno == EINPROGRESS) {
+                    FD_ZERO(&myset);
```

Comment:
What does this code? You probably missed selech between FD_SELECT and FD_ISSET? But use select isn't portable here.